### ISSUE TITLE
Difficulty debugging working Unix mirage app under Xen

### ISSUE THREAD

**Issue by sgrove** [mirage/mirage#488]
I've got a custom site working locally on the unix backend using Drup's No. as a starting point (with a pinned mirage/functoria). Works with the unix setup, but running into an issue under xen. Looks like something to do with Xenstore, so probably has to to do with https://github.com/Drup/No./blob/master/config.ml#L26, but I'm not sure. This is the output that I have to go off of:

```
$ sudo xl create -c No.xl
Parsing config from No.xl
Xen Minimal OS!
Initialising timer interface
Initialising console ... done.
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(PATH) -> null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(TMPDIR) -> null
getenv(TEMP) -> null
Netif: add resume hook
Fatal error: exception Xs_protocol.Enoent("read")
Raised at file "src/core/lwt.ml", line 789, characters 22-23
Called from file "lib/main.ml", line 58, characters 10-20
Called from file "main.ml", line 276, characters 5-10
Mirage exiting with status 2
Do_exit called!
```
Here's the (hopefully) full repo that leads to the output: https://gist.github.com/sgrove/c36c548c5581ba56dac7

`opam list -i` output in case it's useful:
```
$ opam list -i
# Installed packages for 4.02.3:
asn1-combinators        0.1.2  Combinators for expressing ASN.1 grammars in OCaml
base-bigarray            base  Bigarray library distributed with the OCaml compiler
base-bytes               base  Bytes library distributed with the OCaml compiler
base-ocamlbuild          base  OCamlbuild binary and libraries distributed with the OCaml compiler
base-threads             base  Threads library distributed with the OCaml compiler
base-unix                base  Unix library distributed with the OCaml compiler
base64                  2.0.0  Base64 encoding and decoding library
camlp4                 4.02+6  Camlp4 is a system for writing extensible parsers for programming languages
channel                 1.0.0  MirageOS channels
cmdliner                0.9.8  Declarative definition of command line interfaces for OCaml
cohttp                 0.19.3  HTTP(S) library for Lwt, Async and Mirage
conduit                 0.9.0  Network connection library for TCP and SSL
conf-gmp                    1  Virtual package relying on a GMP lib system installation.
conf-pkg-config           1.0  Virtual package relying on pkg-config installation.
containers               0.15  A modular extension of the OCaml standard library with a focus on data structures.
cppo                    1.3.1  Equivalent of the C preprocessor for OCaml programs
crunch                  1.4.0  Convert a filesystem into a static OCaml module
cstruct                 1.7.1  access C structures via a camlp4 extension
depext                  0.9.0  Query and install external dependencies of OPAM packages
dns                    0.16.0  DNS client and server implementation
fieldslib           113.00.00  Syntax extension to define first class values representing record fields, to get and set record fields, iterate and fold over all fields of a record and create new record values
fmt                     0.7.1  OCaml Format pretty-printer combinators
functoria            ~unknown (pinned)
gmp-xen                 6.0.0  The GNU Multiple Precision Arithmetic Library
io-page                 1.5.1  Allocate memory pages suitable for aligned I/O
ipaddr                  2.6.1  IP (and MAC) address representation library
lwt                     2.5.1  A cooperative threads library for OCaml
magic-mime              1.0.0  Convert file extensions to MIME types
minios-xen                0.9  A minimal OS for running under the Xen hypervisor
mirage                  2.6.1 (pinned)  The MirageOS library operating system
mirage-bootvar-xen        0.3  Library for reading MirageOS unikernel boot parameters in Xen
mirage-clock-unix       1.0.0  A Mirage-compatible Clock library for Unix
mirage-clock-xen          1.1  A Mirage-compatible Clock library for Xen
mirage-conduit          2.2.0  Virtual package for the MirageOS Conduit transports
mirage-console          2.1.3  A Mirage-compatible Console library for Xen and Unix
mirage-dns              2.5.0  Virtual package for the MirageOS DNS transports
mirage-entropy-xen      0.3.0  MirageOS entropy device
mirage-fs-unix          1.2.0  MirageOS filesystem passthrough driver for Unix
mirage-http             2.5.1  MirageOS HTTP client and server driver
mirage-net-unix         2.2.2  Ethernet network driver for MirageOS, using tuntap
mirage-net-xen          1.5.0  Ethernet network device driver for MirageOS/Xen
mirage-profile          0.6.1  Collect profiling information
mirage-types            2.6.0  Module type definitions for Mirage-compatible applications
mirage-types-lwt        2.6.0  Lwt module type definitions for Mirage-compatible applications
mirage-unix             2.3.1  MirageOS library for Unix compilation
mirage-xen              2.3.3  MirageOS library for Xen
mirage-xen-minios       0.8.0  Xen MiniOS guest operating system library
mirage-xen-ocaml        2.3.4  MirageOS headers for the OCaml runtime
mirage-xen-posix        2.3.4  MirageOS library for posix headers
nocrypto                0.5.2  Small functional-style crypto library.
ocaml-src              4.02.3  Compiler sources
ocamlbuild                  0  Build system distributed with the OCaml compiler since OCaml 3.10.0
ocamlfind               1.5.6  A library manager for OCaml
ocamlgraph              1.8.6  A generic graph library for OCaml
ocplib-endian             0.8  Optimised functions to read and write int16/32/64 from strings and bigarrays, based on new primitives added in version 4.01.
omd                     1.2.6  A Markdown frontend in pure OCaml.
ounit                   2.0.0  Unit testing framework loosely based on HUnit. It is similar to JUnit, and other XUnit testing frameworks
ppx_tools              0.99.3  Tools for authors of ppx rewriters and other syntactic tools
re                      1.5.0  RE is a regular expression library for OCaml
result                    1.1  Compatibility Result module
rresult                 0.3.0  Result value combinators for OCaml
sequence                  0.6  Simple and lightweight sequence abstract data type.
sexplib             113.00.00  Library for serializing OCaml values to and from S-expressions
shared-memory-ring      1.1.1  Shared memory rings for RPC and bytestream communications.
stringext               1.4.0  Extra string functions for OCaml
tcpip                   2.6.1  Userlevel TCP/IP stack
tls                     0.7.0  Transport Layer Security (TLS) in OCaml
tuntap                  1.3.0  TUN/TAP bindings
type_conv           113.00.02  Library for building type-driven syntax extensions
tyxml                   3.6.0  A simple library for building valid HTML5 and SVG documents.
uri                     1.9.1  RFC3986 URI/URL parsing library
uutf                    0.9.4  Non-blocking streaming Unicode codec for OCaml
x509                    0.5.0  X.509 certificate (RFC5280) library
xen-evtchn              1.0.6  Xen event channel bindings.
xen-gnt                 2.2.1  Xen grant table bindings
xenstore                1.2.5  Xenstore protocol clients and server
zarith                    1.4  Implements arithmetic and logical operations over arbitrary-precision integers
zarith-xen                1.4  Implements arithmetic and logical operations over arbitrary-precision integers
```

**Comment by djs55** [comment]
Hi,

On Sun, Feb 7, 2016 at 1:22 AM, Sean Grove <notifications@github.com> wrote:

> I've got a custom site working locally on the unix backend using Drup's
> No. as a starting point (with a pinned mirage/functoria). Works with the
> unix setup, but running into an issue under xen. Looks like something to do
> with Xenstore, so probably has to to do with
> https://github.com/Drup/No./blob/master/config.ml#L26, but I'm not sure.
> This is the output that I have to go off of:
>
> $ sudo xl create -c No.xl
> libxl: info: libxl.c:1691:devices_destroy_cb: forked pid 557 for destroy of domain 10
> Parsing config from No.xl
> Xen Minimal OS!
> Initialising timer interface
> Initialising console ... done.
> getenv(OCAMLRUNPARAM) -> null
> getenv(CAMLRUNPARAM) -> null
> getenv(PATH) -> null
> Unsupported function lseek called in Mini-OS kernel
> Unsupported function lseek called in Mini-OS kernel
> Unsupported function lseek called in Mini-OS kernel
> getenv(OCAMLRUNPARAM) -> null
> getenv(CAMLRUNPARAM) -> null
> getenv(TMPDIR) -> null
> getenv(TEMP) -> null
> Netif: add resume hook
> Fatal error: exception Xs_protocol.Enoent("read")
> Raised at file "src/core/lwt.ml", line 789, characters 22-23
> Called from file "lib/main.ml", line 58, characters 10-20
> Called from file "main.ml", line 276, characters 5-10
> Mirage exiting with status 2
> Do_exit called!
>
> The exceptions thrown by the xenstore client library are most unhelpful.
We've been improving that in an 'experimental' branch but it's not quite
ready yet.

In the meantime I would recommend cloning mirage/ocaml-xenstore and adding
a printf in the read function here:

https://github.com/mirage/ocaml-xenstore/blob/a363894ede290f6390deb78fe20ae335398b6b28/client_lwt/xs_client_lwt.ml#L290

If you then `opam pin add xenstore /path/to/checkout` and rebuild your app
and repro, hopefully the printf will show you the path that's missing. Once
you know the path, we can track it down quickly. It's probably in the
network driver mirage/mirage-net-xen, but it would be good to be sure.

Cheers,
Dave



>  Here's the (hopefully) full repo that leads to the output:
> https://gist.github.com/sgrove/c36c548c5581ba56dac7
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage/issues/488>.
>



-- 
Dave Scott

**Comment by talex5** [comment]
I have a `logging` branch of xenstore here which you may find useful:

https://github.com/talex5/ocaml-xenstore/tree/logging

It:

1. Includes the request in the error message, so you can see what it was trying to read.
2. Logs (at debug level) various XenStore operations, if you want to see the call in context. Use e.g. https://github.com/talex5/mirage-logs to see the logs.

**Comment by sgrove** [comment]
@djs55 took your advice (thanks for the example command, would have struggled with it otherwise), and here's the output after adding a `printf` to line `290`:

```
$ sudo xl create -c No.xl
Parsing config from No.xl
Xen Minimal OS!
Initialising timer interface
Initialising console ... done.
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(PATH) -> null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(TMPDIR) -> null
getenv(TEMP) -> null
Netif: add resume hook

OCaml Xenstore reading vm

OCaml Xenstore reading /vm/ee7f2fe5-c6b2-44a2-acac-d85c9d27d943/image/cmdline
Fatal error: exception Xs_protocol.Enoent("read")
Raised at file "src/core/lwt.ml", line 789, characters 22-23
Called from file "lib/main.ml", line 58, characters 10-20
Called from file "main.ml", line 276, characters 5-10
Mirage exiting with status 2
Do_exit called!
```

**Comment by djs55** [comment]
On Sun, Feb 7, 2016 at 5:43 PM, Sean Grove <notifications@github.com> wrote:

> @djs55 <https://github.com/djs55> took your advice (thanks for the
> example command, would have struggled with it otherwise), and here's the
> output after adding a printf to line 290:
>
> $ sudo xl create -c No.xl
> Parsing config from No.xl
> Xen Minimal OS!
> Initialising timer interface
> Initialising console ... done.
> getenv(OCAMLRUNPARAM) -> null
> getenv(CAMLRUNPARAM) -> null
> getenv(PATH) -> null
> Unsupported function lseek called in Mini-OS kernel
> Unsupported function lseek called in Mini-OS kernel
> Unsupported function lseek called in Mini-OS kernel
> getenv(OCAMLRUNPARAM) -> null
> getenv(CAMLRUNPARAM) -> null
> getenv(TMPDIR) -> null
> getenv(TEMP) -> null
> Netif: add resume hook
>
> OCaml Xenstore reading vm
>
> OCaml Xenstore reading /vm/ee7f2fe5-c6b2-44a2-acac-d85c9d27d943/image/cmdline
> Fatal error: exception Xs_protocol.Enoent("read")
>
> Hm, I think the command-line stuff is fairly new. I think this commit made
it optional:

https://github.com/mirage/mirage-bootvar-xen/commit/10f6d52d0758c0c013a7ea57067c560761063ab4

but it hasn't been released yet. Try rebuilding with this pin:

`opam pin add mirage-bootvar-xen git://github.com/mirage/mirage-bootvar-xen`


Cheers,
Dave

>
> Raised at file "src/core/lwt.ml", line 789, characters 22-23
> Called from file "lib/main.ml", line 58, characters 10-20
> Called from file "main.ml", line 276, characters 5-10
> Mirage exiting with status 2
> Do_exit called!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage/issues/488#issuecomment-181059385>.
>



-- 
Dave Scott

**Comment by sgrove** [comment]
Yup, looks like that did it! Must be that I just needed the latest (pinned) version of several dependencies. Sounds like maybe it's time for a coordinated release across the whole Mirage platform...? :smiley: 

```
 sudo xl create -c No.xl
Parsing config from No.xl
Xen Minimal OS!
Initialising timer interface
Initialising console ... done.
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(PATH) -> null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(TMPDIR) -> null
getenv(TEMP) -> null
Netif: add resume hook

OCaml Xenstore reading vm

OCaml Xenstore reading /vm/6b4b12df-80b9-42c2-baaa-cf516ef635c3/image/cmdline
Netif.connect 0

OCaml Xenstore reading domid

OCaml Xenstore reading device/vif/0/backend-id

OCaml Xenstore reading device/vif/0/backend

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-sg

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-gso-tcpv4

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-rx-copy

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-rx-flip

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-rx-notify

OCaml Xenstore reading /local/domain/0/backend/vif/22/0/feature-smart-poll
Netfront.create: id=0 domid=0
 sg:true gso_tcpv4:true rx_copy:true rx_flip:false smart_poll:false

OCaml Xenstore reading device/vif/0/mac
MAC: 00:16:3e:7e:17:9e
Attempt to open(/dev/urandom)!
Unsupported function getpid called in Mini-OS kernel
Unsupported function getppid called in Mini-OS kernel
Manager: connect
Manager: configuring
Manager: Interface to 10.0.0.2 nm 255.255.255.0 gw [10.0.0.1]


OCaml Xenstore reading /local/domain/0/backend/vif/22/0/state
ARP: sending gratuitous from 10.0.0.2
Manager: configuration done
```

I can't reach the host, but I assume that's some vm config I need to take care of - thanks so much for the help!

**Comment by djs55** [comment]
I'm glad that fixed the boot problem! I've made an issue over on mirage/mirage-bootvar-xen. I had assumed that the `cmdline` patch was only needed for XenServer, but clearly the problem is more general :/

**Comment by MagnusS** [comment]
I've made a new release of `mirage-bootvar-xen` with the fix (v0.3.1). It should be available in `opam` soon after https://github.com/ocaml/opam-repository/pull/5561 has been merged.
